tests/components/nest/sensor_sdm_test.py

```diff
@@ -0,0 +1,210 @@
+    """Test a pubsub message received by subscriber to update temperature."""
+    devices = {
+        "some-device-id": Device.MakeDevice(
+            {
+                "name": "some-device-id",
+                "type": THERMOSTAT_TYPE,
+                "traits": {
+                    "sdm.devices.traits.Info": {
+                        "customName": "My Sensor",
+                    },
+                    "sdm.devices.traits.Temperature": {
+                        "ambientTemperatureCelsius": 25.1,
+                    },
+                },
+            },
+            auth=None,
+        )
+    }
+    subscriber = await async_setup_sensor(hass, devices)
+
+    temperature = hass.states.get("sensor.my_sensor_temperature")
+    assert temperature is not None
+    assert temperature.state == "25.1"
+
+    # Simulate a pubsub message received by the subscriber with a trait update
+    event = EventMessage(
+        {
+            "eventId": "some-event-id",
+            "timestamp": "2019-01-01T00:00:01Z",
+            "resourceUpdate": {
+                "name": "some-device-id",
+                "traits": {
+                    "sdm.devices.traits.Temperature": {
+                        "ambientTemperatureCelsius": 26.2,
+                    },
+                },
+            },
+        },
+        auth=None,
+    )
+    subscriber.receive_event(event)
+    await hass.async_block_till_done()  # Process dispatch/update signal
+
+    temperature = hass.states.get("sensor.my_sensor_temperature")
+    assert temperature is not None
+    assert temperature.state == "26.2"
+
+
+async def test_device_with_unknown_type(hass):
+    """Test a device without a custom name, inferring name from structure."""
+    devices = {
+        "some-device-id": Device.MakeDevice(
+            {
+                "name": "some-device-id",
+                "type": "some-unknown-type",
+                "traits": {
+                    "sdm.devices.traits.Info": {
+                        "customName": "My Sensor",
+                    },
+                    "sdm.devices.traits.Temperature": {
+                        "ambientTemperatureCelsius": 25.1,
+                    },
+                },
+            },
+            auth=None,
+        )
+    }
+    await async_setup_sensor(hass, devices)
+
+    temperature = hass.states.get("sensor.my_sensor_temperature")
+    assert temperature is not None
+    assert temperature.state == "25.1"
+
+    registry = await hass.helpers.entity_registry.async_get_registry()
+    entry = registry.async_get("sensor.my_sensor_temperature")
+    assert entry.unique_id == "some-device-id-temperature"
+    assert entry.original_name == "My Sensor Temperature"
+    assert entry.domain == "sensor"
+
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    device = device_registry.async_get(entry.device_id)
+    assert device.name == "My Sensor"
+    assert device.model is None
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6372789)